text/3716-target-modifiers.md

```diff
@@ -0,0 +1,517 @@
+the ABI used in the kernel. Often, this is done conditionally depending on a
+configuration option. A few examples:
+
+* When using `CONFIG_SHADOW_CALLSTACK` the x18 register must be reserved in the
+  ABI with `-Zfixed-x18`.
+* The `-Ctarget-feature=-neon` flag is used to prevent use of floating points
+  on arm.
+* On 32-bit x86, `-Zreg-struct-return` and `-Zregparm=3` are used.
+* When using `CONFIG_CFI_CLANG` the kCFI sanitizer is enabled with
+  `-Zsanitizer=kcfi`. Unlike most other sanitizers, this sanitizer is used in
+  production.
+* In several different cases, `-Zpatchable-function-entry` is used to add nops
+  before or after the function entrypoint. When mixed with `-Zsanitizer=kcfi`
+  this causes special considerations as kcfi works by placing a tag before the
+  function entrypoint.
+* To support `MITIGATION_RETPOLINE` and `MITIGATION_SLS`, `target.json` is used
+  on x86.
+
+We expect there to be more examples in the future.
+
+## Sanitizers
+
+There is [an ongoing effort to stabilize some of the sanitizers][issue123615].
+However, this effort explicitly aims to stabilize sanitizers that can be used
+without rebuilding the stdlib. With this RFC, that is no longer a blocker as
+the remaining sanitizers can be classified as a target modifier and then
+stabilized.
+
+[issue123615]: https://github.com/rust-lang/rust/issues/123615
+
+## Embedded Targets
+
+Currently, embedded platforms such as `thumb*` or `rv*` use separate targets
+for configuration with significant ABI changes. For `thumb*` targets, this is
+currently limited to Hard- vs Soft-Float which can cause issues when linked.
+However for RISC-V targets, the F (32-bit hardware float), D (64-bit hardware
+float), and Q (128-bit hardware float) extensions all can [potentially change
+the ABI][riscv-float], which would increase the number of required targets. The
+[E extension][riscv-e] may also change the ABI by limiting the number of
+registers used by the I (integer operations) extension.
+
+[riscv-float]: https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-cc.adoc#named-abis
+[riscv-e]: https://github.com/riscv/riscv-isa-manual/blob/main/src/rv32e.adoc
+
+## Existing -C flags that are unsound
+
+It has recently been discovered that several existing `-C` flags modify the
+ABI, making them unsound. Examples:
+
+* [`-Csoft-float`](https://github.com/rust-lang/rust/issues/129893)
+* [`-Ctarget-feature=-neon`](https://github.com/rust-lang/rust/issues/131058)
+* [`-Clinker-plugin-lto`](https://github.com/rust-lang/rust/issues/127979)
+* [`-Cllvm-args`](https://github.com/rust-lang/rust/issues/131800#issuecomment-2418595757)
+* Possibly `-Ccode-model` and `-Crelocation-model`
+
+This problem is a new discovery and it's still not clear how to solve it.
+Target modifiers will not solve all of the flags; [some flags are just
+unfixable and need to be removed][issue130968]. But I expect that target
+modifiers will be the solution for some of these flags.
```

Comment:
Is there an issue listing flags that we will want to make target modifiers? Might be good to collect future work in one place if that doesn't exist yet.

Reply:
There are some individual issues (e.g., [rust#129893][1] and [rust#131058][2]) for flags that are unsound due to ABI issues, but I don't think we have a combined issue tracking all of them.

[1]: https://github.com/rust-lang/rust/issues/129893
[2]: https://github.com/rust-lang/rust/issues/131058

Reply:
There is https://github.com/rust-lang/rust/issues/131837